bids/analysis/transformations/compute.py

```diff
@@ -35,6 +36,25 @@ def _transform(self, var, model='spm', derivative=False, dispersion=False,
 
         if isinstance(var, SparseRunVariable):
             sr = self.collection.sampling_rate
+            # Resolve diferences in TR and TA to milliseconds
+            trs = {ri.tr for ri in var.run_info}
+            tas = {ri.ta for ri in var.run_info}
+            assert len(trs) == 1
+            assert len(tas) == 1
+            TR = int(np.round(1000. * trs.pop()))
+            TA = int(np.round(1000. * tas.pop()))
+            SR = int(np.round(1000. / sr))
+            if TA is None or TA < TR:
+                # Use a unit that fits an whole number of times into both
+                # the interscan interval (TR) and the integration window (TA)
+                dt = gcd(TR, TA)
+                if dt > SR:
```

Comment:
I think I'm probably missing something, but if we have to give up on a value that neatly bins `TR` and `TA`, why not just use `SR` at that point? E.g., if dt = 50 and SR = 10, then this will give us  dt = 10, so we're at the SR anyway. If dt = 50 and SR = 17, we get dt = 25, which both prevents neat binning and is bigger than we wanted. Is the idea that at least every `dt//SR`^th bin will align nicely this way?

Reply:
25 also permits neat binning...

Reply:
Oh, duh. The inversion tripped me up. That is indeed the thing I was missing. :)

Reply:
That said, I'm still not sure we shouldn't just use the `SR`. Since the user could have set the `sr` on the collection manually, the compromise solution here risks explicitly ignoring their intentions. This isn't an entirely benign decision, because in subsequent transformations, if a mismatch in sampling rates is detected, that will trigger a resampling in at least one variable, when potentially none might have been needed. They could also have code that depends on knowing what the SR is.

I think maybe we need to come up with a general policy about this, because it crops up in other places too (e.g., the issue about what to do if event files contain very short durations). I.e., the general question is, in a situation where the user appears to be picking sampling rates that are objectively inefficient or result in loss of information, should we be paternalistic and fix it for them, or just let them know they can probably do better?

Reply:
One thing we could also do, though it's a bit annoying, is internally maintain a flag that indicates whether or not the user has explicitly overridden the default sampling rate. If `SR=10` purely because that's the default, overriding it seems fine. If the user explicitly set it themselves, maybe we don't want to pick a different value for them.

Reply:
It feels like we're trying to thread a needle where we make smart default choices based on our expertise on the one hand, and on the other, loading a gun, cocking it and pointing it at the foot of anyone who looks like they might know what they're doing...

Currently we reserve the footgun for people who use `ToDense` to get themselves a sampling interval that is relatively prime to `gcd(TR, TA)`. If someone knows enough to manipulate the default sampling rate, they can learn that `Convolve` with a sparse acquisition paradigm will choose a slightly different interval than with a continuous acquisition paradigm unless the variable is already dense.

But perhaps there's a better general approach here? I admit that I chose this simply because building a boxcar function and taking a mean is easier than learning how to do this with `interp1d`.

Reply:
The general issue potentially affects anyone who sets the `sampling_rate` explicitly in `BIDSLayout.get_collections` or `load_variables`. A sophisticated user could certainly know enough about their dataset to think "I'll use a sampling rate of 50, because that's high enough to account for my short-duration events, but still manageable computationally". Thereafter, I think that user could reasonably expect that if they call `ToDense` or `Convolve` without explicitly specifying a different sampling rate, the resulting variable will have an effective sampling rate of 50.

I think making decisions for the user internally is fine if there's an explicit argument in the transformation that implies as much. I.e., if `ToDense` has default `sampling_rate='auto'`, then the docstring can just explain that we will act intelligently unless a numeric value is passed, and the result may not equal what's currently set in the collection. That seems fine. The issue is that in this case, there's no provision to specify the sampling rate in `Convolve`. I was against doing that because I don't think the spec should have that argument, but I guess we can make it a pybids-only argument, if only for the sake of making it clear to a sophisticated user what's going on.